include/dcgp/problems/symbolic_regression.hpp

```diff
@@ -75,33 +112,98 @@ class symbolic_regression
+            // 1 - We extract the integer part and represent it as an unsigned vector to set the CGP expression.
+            std::vector<unsigned> xu(x.size() - m_n_eph);
+            std::transform(x.data() + m_n_eph, x.data() + x.size(), xu.begin(),
+                           [](double a) { return boost::numeric_cast<unsigned>(a); });
+            m_cgp.set(xu);
+            // 2 - We set the floating point part as ephemeral constants.
+            std::vector<double> eph_val(x.data(), x.data() + m_n_eph);
+            m_cgp.set_eph_val(eph_val);
+            // 3 - We compute the MSE loss splitting the data in n batches.
+            // TODO: make this work also when m_parallel_batches does not divide exactly the data size.
+            retval[0] = m_cgp.loss(m_points, m_labels, "MSE", m_parallel_batches);
+            return retval;
+        }
+    }
+
+    /// Gradient computation
+    /**
+     * Computes the gradient for the continuous part of this UDP
+     *
+     * @param x the decision vector.
+     *
+     * @return the gradient of \p x.
+     */
+    pagmo::vector_double gradient(const pagmo::vector_double &x) const
+    {
+        std::vector<double> retval(m_n_eph, 0);
         // The chromosome has a floating point part (the ephemeral constants) and an integer part (the encoded CGP).
         // 1 - We extract the integer part and represent it as an unsigned vector to set the CGP expression.
         std::vector<unsigned> xu(x.size() - m_n_eph);
-        std::transform(x.data() + m_n_eph, x.data() + x.size(), xu.data(),
+        std::transform(x.data() + m_n_eph, x.data() + x.size(), xu.begin(),
                        [](double a) { return boost::numeric_cast<unsigned>(a); });
-        m_cgp.set(xu);
-        // 2 - We set the floating point part as ephemeral constants.
-        std::vector<double> eph_val(x.data(), x.data() + m_n_eph);
-        m_cgp.set_eph_val(eph_val);
+        m_dcgp.set(xu);
+        // 2 - We use the floating point part of the chromosome to set ephemeral constants.
+        std::vector<audi::gdual_d> eph_val;
+        for (decltype(m_n_eph) i = 0u; i < m_n_eph; ++i) {
+            eph_val.emplace_back(x[i], m_dcgp.get_eph_symb()[i], 1u); // Only first derivative is needed
+        }
+        m_dcgp.set_eph_val(eph_val);
         // 3 - We compute the MSE loss splitting the data in n batches.
         // TODO: make this work also when m_parallel_batches does not divide exactly the data size.
-        retval[0] = m_cgp.loss(m_points, m_labels, "MSE", m_parallel_batches);
+        auto loss = m_dcgp.loss(m_dpoints, m_dlabels, "MSE", m_parallel_batches);
+        // since we have also computed the loss value, we store it in a cache so that fitness can be called and
+        // not cause reavaluation of a cgp.
+        m_cache = decltype(m_cache){x, pagmo::vector_double{loss.constant_cf()}};
```

Comment:
Here it's better to assign directly the members of the pair rather than creating a new pair and then move-assign it.

```c++
m_cache.first = x;
m_cache.second = pagmo::vector_double{loss.constant_cf()};
```

Even better would be to avoid altogether the creation of a temporary ``vector_double``. Perhaps something like

```c++
m_cache.second.resize(1);
m_cache.second[0] = loss.constant_cf();
```

Reply:
yeps, implemented thanks!